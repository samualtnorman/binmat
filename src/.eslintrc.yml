extends: "@samual/eslint-config/typescript.yml"

rules:
  "@typescript-eslint/brace-style": off
