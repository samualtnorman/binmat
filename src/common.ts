export type State = {
	attackerStacks: [ Card[], Card[], Card[], Card[], Card[], Card[] ]
	defenderStacks: [ DefenderStack, DefenderStack, DefenderStack, DefenderStack, DefenderStack, DefenderStack ]
	laneDecks: [ Card[], Card[], Card[], Card[], Card[], Card[] ]
	laneDiscardPiles: [ Card[], Card[], Card[], Card[], Card[], Card[] ]
	attackerDeck: Card[]
	attackerDiscardPile: Card[]
	attackerHand: Card[]
	defenderHand: Card[]
	turn: number
	// TODO think about renaming or removing `turns`
	turns: number
	attackerPassedLastTurn: boolean
	defenderPassedLastTurn: boolean
}

export type Card = `${CardValue}${CardSuit}`
export type CardValue = CardNumber | CardModifier
export type CardNumber = "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "a"

export const enum CardModifier {
	Trap = `@`, Wild = `*`, Bounce = `?`, Break = `>`
}

export const enum CardSuit {
	Form = `&`, Kin = `%`, Data = `+`, Chaos = `!`, Void = `^`, Choice = `#`
}

export type DefenderStack = { cards: Card[], isFaceUp: boolean }

export type Move =
	{ kind: MoveKind.Draw, deck: Lane | AttackerDeck } |
	{ kind: MoveKind.Play | MoveKind.PlayFaceUp, card: Card | CardValue, lane: Lane } |
	{ kind: MoveKind.Combat, lane: Lane } |
	{ kind: MoveKind.Discard, card: Card | CardValue, discardPile: Lane | AttackerDiscardPile } |
	{ kind: MoveKind.Pass }

export const enum MoveKind {
	Draw, Play, PlayFaceUp, Combat, Discard, Pass
}

export type Lane = 0 | 2 | 1 | 3 | 4 | 5

export const AttackerDeck = 6
export type AttackerDeck = typeof AttackerDeck

export const AttackerDiscardPile = 6
export type AttackerDiscardPile = typeof AttackerDiscardPile

export const enum StatusCode {
	Ok, DefenderWin, AttackerWin, MadeMoveOnFinishedGame, DefenderDrewFromAttackerDeck, AttackerDrewFromBlockedLane,
	PlayedUnownedCard, PlayedBreakToEmptyStack, DefenderInitiatedCombat, AttackerInitiatedCombatWithEmptyStack,
	DiscardedToOpponentDiscardPile, AttackerDiscardedToEmptyDiscardAndDeck, AttackerDrewFromEmptyDiscardAndDeck,
	PlayedCardFacedWrongWay, DefenderPlayedFaceUpBreakToStackWithBreak
}

export const StatusCodeMessages: Record<StatusCode, string> = [
	`okay`,
	`defender won`,
	`attacker won`,
	`tried to make a move on a finished game`,
	`the defender tried to draw from the attacker's deck`,
	`the attacker tried to drawn from a blocked lane deck`,
	`tried to play a card not in hand`,
	`tried to play a break card to an empty stack`,
	`the defender tried to initiate combat`,
	`the attacker tried to initiate combat with an empty attacker stack`,
	`tried to discard to opponent's discard pile`,
	`attacker tried to discard when the attacker discard pile and attacker deck are empty`,
	`attacker tried to draw from attacker deck when it and the attacker discard pile are empty`,
	`tried to play card faced incorrectly`,
	`the defender tried to play a face up break to a stack that already contains a break card`
]

export const enum Role {
	Defender = 1, Attacker
}
