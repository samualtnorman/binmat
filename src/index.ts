export { doMove } from "./doMove"
export { makeState } from "./makeState"
export { parseMove } from "./parseMove"
export { MoveKind, Role, StatusCode, StatusCodeMessages, type Move, type State } from "./shared"
export { simulateGame } from "./simulateGame"
