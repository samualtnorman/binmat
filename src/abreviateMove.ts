import { MoveTag, type Move, type MoveString } from "./common"

export function abreviateMove(move: Move): MoveString {
	switch (move.tag) {
		case MoveTag.Draw:
			return `d${move.deck == 6 ? `a` : move.deck}`

		case MoveTag.Combat:
			return `c${move.lane}`

		case MoveTag.Play:
			return `p${move.card}${move.lane}`

		case MoveTag.PlayFaceUp:
			return `u${move.card}${move.lane}`

		case MoveTag.Discard:
			return `x${move.card}${move.discardPile == 6 ? `a` : move.discardPile}`

		case MoveTag.Pass:
			return `--`
	}
}
