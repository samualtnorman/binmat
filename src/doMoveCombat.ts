import type { CombatData } from "./doCombat"
import { doCombat } from "./doCombat"
import type { Lane, State } from "./common"
import { Role, StatusCode } from "./common"

export function doMoveCombat(state: State, lane: Lane): (
	{ status: StatusCode.Okay | StatusCode.DefenderWin | StatusCode.AttackerWin } & CombatData
) | {
	status:
		StatusCode.MadeMoveOnFinishedGame |
		StatusCode.DefenderInitiatedCombat |
		StatusCode.AttackerInitiatedCombatWithEmptyStack
} {
	if (state.turn >= state.maxTurns)
		return { status: StatusCode.MadeMoveOnFinishedGame }

	const roleTurn: Role = (state.turn % 2) + 1

	if (roleTurn == Role.Defender)
		return { status: StatusCode.DefenderInitiatedCombat }

	if (!state.attackerStacks[lane].length)
		return { status: StatusCode.AttackerInitiatedCombatWithEmptyStack }

	const combatResult = doCombat(state, lane)

	if (combatResult.status == StatusCode.AttackerWin)
		return combatResult

	state.turn++

	if (state.turn == state.maxTurns)
		return { ...combatResult, status: StatusCode.DefenderWin }

	return combatResult
}
