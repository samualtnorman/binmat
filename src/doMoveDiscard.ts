import { shuffle } from "@samual/lib/shuffle"
import type { Card, CardValue, Lane, State } from "./common"
import { AttackerDiscardPile, Role, StatusCode } from "./common"

export function doMoveDiscard(state: State, card: Card | CardValue, discardPile: Lane | AttackerDiscardPile): {
	status: StatusCode.Ok | StatusCode.DefenderWin
	cardDiscarded: Card
	cardsDrawn: [ Card, Card ] | undefined
} | {
	status:
		StatusCode.MadeMoveOnFinishedGame |
		StatusCode.PlayedUnownedCard |
		StatusCode.DiscardedToOpponentDiscardPile |
		StatusCode.AttackerDiscardedToEmptyDiscardAndDeck
} {
	if (state.turn >= state.maxTurns)
		return { status: StatusCode.MadeMoveOnFinishedGame }

	const roleTurn: Role = (state.turn % 2) + 1
	let cardDiscarded
	let cardsDrawn: [ Card, Card ] | undefined

	if (roleTurn == Role.Defender) {
		const index = card.length == 2
			? state.defenderHand.indexOf(card as Card)
			: state.defenderHand.findIndex(([ value ]) => value == card)

		if (index == -1)
			return { status: StatusCode.PlayedUnownedCard }

		if (discardPile == AttackerDiscardPile)
			return { status: StatusCode.DiscardedToOpponentDiscardPile }

		cardDiscarded = state.defenderHand.splice(index, 1)[0]!
		state.laneDiscardPiles[discardPile].push(cardDiscarded)
	} else /* attacker turn */ {
		const index = card.length == 2
			? state.attackerHand.indexOf(card as Card)
			: state.attackerHand.findIndex(([ value ]) => value == card)

		if (index == -1)
			return { status: StatusCode.PlayedUnownedCard }

		if (discardPile != AttackerDiscardPile)
			return { status: StatusCode.DiscardedToOpponentDiscardPile }

		if (!state.attackerDiscardPile.length && !state.attackerDeck.length)
			return { status: StatusCode.AttackerDiscardedToEmptyDiscardAndDeck }

		cardDiscarded = state.attackerHand.splice(index, 1)[0]!
		state.attackerDiscardPile.push(cardDiscarded)

		if (state.attackerDeck.length == 1) {
			const firstCardDrawn = state.attackerDeck.pop()!

			state.attackerHand.push(firstCardDrawn)
			state.attackerDeck.push(...shuffle(state.attackerDiscardPile.splice(0)))
			cardsDrawn = [ firstCardDrawn, state.attackerDeck.pop()! ]
			state.attackerHand.push(cardsDrawn[1])
		} else {
			if (!state.attackerDeck.length)
				state.attackerDeck.push(...shuffle(state.attackerDiscardPile.splice(0)))

			cardsDrawn = [ state.attackerDeck.pop()!, state.attackerDeck.pop()! ]
			state.attackerHand.push(cardsDrawn[0], cardsDrawn[1])
		}
	}

	state.turn++

	if (state.turn == state.maxTurns)
		return { status: StatusCode.DefenderWin, cardDiscarded, cardsDrawn }

	return { status: StatusCode.Ok, cardDiscarded, cardsDrawn }
}
