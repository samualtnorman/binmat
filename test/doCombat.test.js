import { expect, test, vi } from "vitest"
import { doCombat } from "../src/doCombat"
import { StatusCode } from "../src/common"

vi.mock("@samual/lib/shuffle", () =>
	/** @satisfies {typeof import("@samual/lib/shuffle")} */ ({ shuffle: array => array })
)

test("trap (defender)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "@!", "@#" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.Okay,
		attackerStack: [ "8!" ],
		defenderStack: [ "@!", "@#" ],
		attackerAttackPower: 0,
		defenderAttackPower: 0,
		damageValue: 0,
		cardsDrawn: [],
		attackerBouncesDiscarded: [],
		attackerCardsTrapped: [ "8!" ],
		attackerStackDiscarded: [],
		defenderBouncesDiscarded: [],
		defenderCardsTrapped: [],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: []
	})

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "@!", "@#" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [ "8!" ], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("wild (defender)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "5!", "*!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.Okay,
		attackerStack: [ "8!" ],
		defenderStack: [ "5!", "*!" ],
		attackerAttackPower: 3,
		defenderAttackPower: 3,
		damageValue: 1,
		cardsDrawn: [],
		attackerBouncesDiscarded: [],
		attackerCardsTrapped: [],
		attackerStackDiscarded: [ "8!" ],
		defenderBouncesDiscarded: [],
		defenderCardsTrapped: [],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: [ "*!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "5!" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [ "*!", "8!" ],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("bounce (defender)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "?!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toMatchSnapshot()

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [ "?!", "8!" ],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("break (defender)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "2!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "3!", "3#", ">!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.Okay,
		attackerStack: [ "2!" ],
		defenderStack: [ "3!", "3#", ">!" ],
		attackerAttackPower: 1,
		defenderAttackPower: 0,
		damageValue: 3,
		cardsDrawn: [],
		attackerBouncesDiscarded: [],
		attackerCardsTrapped: [],
		attackerStackDiscarded: [ "2!" ],
		defenderBouncesDiscarded: [],
		defenderCardsTrapped: [],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: [ ">!", "3#", "3!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [ ">!", "3#", "3!", "2!" ],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("10 (defender)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "a!", "6!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.Okay,
		attackerStack: [ "8!" ],
		defenderStack: [ "a!", "6!" ],
		attackerAttackPower: 3,
		defenderAttackPower: 4,
		damageValue: 0,
		cardsDrawn: [],
		attackerBouncesDiscarded: [],
		attackerCardsTrapped: [],
		attackerStackDiscarded: [ "8!" ],
		defenderBouncesDiscarded: [],
		defenderCardsTrapped: [],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: []
	})

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "a!", "6!" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [ "8!" ], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 0,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("trap (attacker)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "@!", "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "?!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.AttackerWin,
		attackerStack: [ "@!", "8!" ],
		defenderStack: [ "?!" ],
		attackerAttackPower: 3,
		defenderAttackPower: 0,
		damageValue: 4,
		cardsDrawn: [],
		attackerCardsTrapped: [],
		defenderCardsTrapped: [ "?!" ],
		attackerBouncesDiscarded: [],
		defenderBouncesDiscarded: [],
		attackerStackDiscarded: [ "@!", "8!" ],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: []
	})

	expect(state).toEqual({
		attackerStacks: [ [ "@!", "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [ "?!" ],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("wild (attacker)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "5!", "*!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "4!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.AttackerWin,
		attackerStack: [ "5!", "*!" ],
		defenderStack: [ "4!" ],
		attackerAttackPower: 3,
		defenderAttackPower: 2,
		damageValue: 2,
		cardsDrawn: [],
		attackerCardsTrapped: [],
		defenderCardsTrapped: [],
		attackerBouncesDiscarded: [],
		defenderBouncesDiscarded: [],
		attackerStackDiscarded: [ "5!", "*!" ],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: [ "4!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [ "5!", "*!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "4!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("bounce (attacker)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "?!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "4!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toMatchSnapshot()

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "4!" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [ "?!" ], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("break (attacker)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "2!", ">!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "3!", "3#", "3%" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.Okay,
		attackerStack: [ "2!", ">!" ],
		defenderStack: [ "3!", "3#", "3%" ],
		attackerAttackPower: 1,
		defenderAttackPower: 0,
		damageValue: 3,
		cardsDrawn: [],
		attackerBouncesDiscarded: [],
		attackerCardsTrapped: [],
		attackerStackDiscarded: [ "2!", ">!" ],
		defenderBouncesDiscarded: [],
		defenderCardsTrapped: [],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: [ "3%", "3#", "3!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [ "3%", "3#", "3!", "2!", ">!" ],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("10 (attacker)", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "a!", "6!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "8!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.AttackerWin,
		attackerStack: [ "a!", "6!" ],
		defenderStack: [ "8!" ],
		attackerAttackPower: 4,
		defenderAttackPower: 3,
		damageValue: 2,
		cardsDrawn: [],
		attackerCardsTrapped: [],
		defenderCardsTrapped: [],
		attackerBouncesDiscarded: [],
		defenderBouncesDiscarded: [],
		attackerStackDiscarded: [ "a!", "6!" ],
		defenderStackWasFaceUp: false,
		cardsDrawnToDiscard: [ "8!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [ "a!", "6!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "8!" ], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})

test("shuffling lane discard pile into lane deck", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [ "3!" ], [], [], [], [], [] ],
		laneDiscardPiles: [ [ "3#", "3%", "3&", "3+", "3^" ], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toMatchSnapshot()
	expect(state).toMatchSnapshot()
})

test("defender stack already face up", () => {
	/** @type {import("../src/common").State} */
	const state = {
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "@!" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	}

	expect(doCombat(state, 0)).toEqual({
		status: StatusCode.AttackerWin,
		attackerStack: [ "8!" ],
		defenderStack: [ "@!" ],
		attackerAttackPower: 3,
		defenderAttackPower: 0,
		damageValue: 4,
		cardsDrawn: [],
		attackerCardsTrapped: [],
		defenderCardsTrapped: [],
		attackerBouncesDiscarded: [],
		defenderBouncesDiscarded: [],
		attackerStackDiscarded: [ "8!" ],
		defenderStackWasFaceUp: true,
		cardsDrawnToDiscard: [ "@!" ]
	})

	expect(state).toEqual({
		attackerStacks: [ [ "8!" ], [], [], [], [], [] ],
		defenderStacks: [
			{ cards: [ "@!" ], isFaceUp: true }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false },
			{ cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }, { cards: [], isFaceUp: false }
		],
		laneDecks: [ [], [], [], [], [], [] ],
		laneDiscardPiles: [ [], [], [], [], [], [] ],
		attackerDeck: [],
		attackerDiscardPile: [],
		attackerHand: [],
		defenderHand: [],
		turn: 1,
		maxTurns: 110,
		attackerPassedLastTurn: false,
		defenderPassedLastTurn: false
	})
})
