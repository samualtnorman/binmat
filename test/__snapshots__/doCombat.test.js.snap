// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bounce (attacker) 1`] = `
{
  "attackerAttackPower": 0,
  "attackerBouncesDiscarded": [
    "?!",
  ],
  "attackerCardsTrapped": [],
  "attackerStack": [
    "?!",
  ],
  "attackerStackDiscarded": [],
  "cardsDrawn": [],
  "cardsDrawnToDiscard": [],
  "damageValue": 0,
  "defenderAttackPower": 0,
  "defenderBouncesDiscarded": [],
  "defenderCardsTrapped": [],
  "defenderStack": [
    "4!",
  ],
  "defenderStackWasFaceUp": false,
  "status": 1,
}
`;

exports[`bounce (defender) 1`] = `
{
  "attackerAttackPower": 0,
  "attackerBouncesDiscarded": [],
  "attackerCardsTrapped": [],
  "attackerStack": [
    "8!",
  ],
  "attackerStackDiscarded": [
    "8!",
  ],
  "cardsDrawn": [],
  "cardsDrawnToDiscard": [],
  "damageValue": 0,
  "defenderAttackPower": 0,
  "defenderBouncesDiscarded": [
    "?!",
  ],
  "defenderCardsTrapped": [],
  "defenderStack": [
    "?!",
  ],
  "defenderStackWasFaceUp": false,
  "status": 1,
}
`;
